File: android/app/src/test/java/com/enctxt/PatternRendererTest.kt
package com.enctxt

import com.enctxt.core.privacy.IntentClassifier
import com.enctxt.core.privacy.PatternRenderer
import org.junit.Assert.*
import org.junit.Test

class PatternRendererTest {

    private val version = 2

    @Test
    fun testEmptyStringReturnsEmptyString() {
        assertEquals("", PatternRenderer.protect("", version))
    }

    @Test
    fun testIsDeterministic() {
        val input = "Are you coming tonight?"
        assertEquals(PatternRenderer.protect(input, version), PatternRenderer.protect(input, version))
    }

    @Test
    fun testDifferentMessagesProduceDifferentOutput() {
        val a = PatternRenderer.protect("Are you coming tonight?", version)
        val b = PatternRenderer.protect("Meet me at the station", version)
        assertNotEquals(a, b)
    }

    @Test
    fun testNeverContainsOriginalPlaintextWords() {
        val input = "The confidential project codename is Falcon"
        val output = PatternRenderer.protect(input, version).lowercase()
        for (word in input.split(Regex("\\s+"))) {
            val cleaned = word.replace(Regex("[^a-zA-Z]"), "")
            if (cleaned.length > 2) {
                assertFalse(output.contains(cleaned.lowercase()))
            }
        }
    }

    @Test
    fun testEmbedsCorrectIntentSymbol() {
        val input = "Are you coming tonight?"
        val intent = IntentClassifier.classify(input)
        val output = PatternRenderer.protect(input, version)
        assertTrue(output.contains(PatternRenderer.INTENT_SYMBOLS.getValue(intent)))
    }

    @Test
    fun testFollowsFixedGrammar() {
        val output = PatternRenderer.protect("Hello there", version)
        val parts = output.split(" ")
        assertEquals(6, parts.size)
        assertEquals("·", parts[2])
        assertEquals("·", parts[4])
    }
}

File: client/test/patternRenderer.test.ts
import { describe, it, expect } from 'vitest';
import { renderPattern, INTENT_SYMBOLS } from '../src/utils/protectedText/patternRenderer';
import { classifyIntent } from '../src/utils/protectedText/intentClassifier';

const VERSION = 2;

describe('PATTERN rendering strategy (Protected Text v2)', () => {
  it('returns empty string for empty input', () => {
    expect(renderPattern('', VERSION)).toBe('');
  });

  it('is deterministic for the same message', () => {
    const input = 'Are you coming tonight?';
    expect(renderPattern(input, VERSION)).toBe(renderPattern(input, VERSION));
  });

  it('produces different output for different messages (in general)', () => {
    const a = renderPattern('Are you coming tonight?', VERSION);
    const b = renderPattern('Meet me at the station', VERSION);
    expect(a).not.toBe(b);
  });

  it('never contains any word from the original plaintext', () => {
    const input = 'The confidential project codename is Falcon';
    const output = renderPattern(input, VERSION);
    for (const word of input.split(/\s+/)) {
      const cleaned = word.replace(/[^a-zA-Z]/g, '');
      if (cleaned.length > 2) {
        expect(output.toLowerCase()).not.toContain(cleaned.toLowerCase());
      }
    }
  });

  it('embeds the correct intent symbol for the classified category', () => {
    const input = 'Are you coming tonight?';
    const intent = classifyIntent(input);
    const output = renderPattern(input, VERSION);
    expect(output).toContain(INTENT_SYMBOLS[intent]);
  });

  it('follows the fixed PREFIX + TOKEN + SEP + SYMBOL + SEP + TOKEN grammar', () => {
    const output = renderPattern('Hello there', VERSION);
    const parts = output.split(' ');
    // prefix, token, ·, symbol, ·, token = 6 space-separated parts
    expect(parts.length).toBe(6);
    expect(parts[2]).toBe('·');
    expect(parts[4]).toBe('·');
  });
});
